hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/RecordLevelIndexSupport.scala

```diff
@@ -122,47 +123,78 @@ object RecordLevelIndexSupport {
    */
-  def filterQueryWithRecordKey(queryFilter: Expression, recordKeyOpt: Option[String]): Option[(Expression, List[String])] = {
+  def filterQueryWithRecordKey(queryFilter: Expression, recordKeyFields: Array[String]): Option[(Expression, List[String])] = {
+    val isComplexRecordKey = recordKeyFields.length > 1
     queryFilter match {
+      // Handle EqualTo expressions
       case equalToQuery: EqualTo =>
         val attributeLiteralTuple = getAttributeLiteralTuple(equalToQuery.left, equalToQuery.right).orNull
         if (attributeLiteralTuple != null) {
           val attribute = attributeLiteralTuple._1
           val literal = attributeLiteralTuple._2
-          if (attribute != null && attribute.name != null && attributeMatchesRecordKey(attribute.name, recordKeyOpt)) {
-            Option.apply(equalToQuery, List.apply(literal.value.toString))
+          if (attribute != null && attribute.name != null && attributeMatchesRecordKey(attribute.name, recordKeyFields)) {
+            val recordKeyLiteral = getRecordKeyLiteral(attribute, literal, isComplexRecordKey)
+            Option.apply(equalToQuery, List.apply(recordKeyLiteral))
           } else {
             Option.empty
           }
         } else {
           Option.empty
         }
 
+      // Handle In expressions
       case inQuery: In =>
         var validINQuery = true
         inQuery.value match {
           case attribute: AttributeReference =>
-            if (!attributeMatchesRecordKey(attribute.name, recordKeyOpt)) {
+            if (!attributeMatchesRecordKey(attribute.name, recordKeyFields)) {
               validINQuery = false
             }
           case _ => validINQuery = false
         }
         var literals: List[String] = List.empty
         inQuery.list.foreach {
-          case literal: Literal => literals = literals :+ literal.value.toString
+          case literal: Literal => literals = literals :+ getRecordKeyLiteral(inQuery.value.asInstanceOf[AttributeReference], literal, isComplexRecordKey)
           case _ => validINQuery = false
         }
         if (validINQuery) {
           Option.apply(inQuery, literals)
         } else {
           Option.empty
         }
+
+      // Handle And expression (composite filter)
+      case andQuery: And =>
+        val leftResult = filterQueryWithRecordKey(andQuery.left, recordKeyFields)
+        val rightResult = filterQueryWithRecordKey(andQuery.right, recordKeyFields)
+
+        // If both left and right filters are valid, concatenate their results
+        (leftResult, rightResult) match {
+          case (Some((leftExp, leftKeys)), Some((rightExp, rightKeys))) =>
+            // Return concatenated expressions and record keys
+            Option.apply(And(leftExp, rightExp), leftKeys ++ rightKeys)
+          case _ => Option.empty
+        }
+
+      // Handle Or expression (for completeness, though usually not used for record key filtering)
+      case orQuery: Or =>
```

Comment:
I guess the `OR` does not work as expected.

Reply:
Yeah..removed the OR handling for now. We can add it in a separate jira.